### DISCUSSION TITLE
Unknown keywords annotation

### DISCUSSION THREAD

**Discussion by gregsdennis** [json-schema-org/community#512] [Specification]
The primary argument against supporting unknown keywords is that, because they're just skipped (or their values collected as annotations), there isn't a good way to catch mis-typed keywords.  That leads to unexpected validation results, where a keyword is expected to perform some validation but doesn't because what should have been a keyword wasn't recognized as one.

The proposed solutions to this were to

- put `additionalProperties: false` in the meta-schema
- disallow unknown keywords entirely

We've opted for the second approach with a little bit of the first (we'll have one open and one closed meta-schema).

But what if we just require an annotation that is a list of the unknown keywords?

```jsonc
{
  "valid": true,
  "evaluationPath": "",
  "schemaLocation": "https://json-everything.net/ac994a0348#",
  "instanceLocation": "",
  "annotations": {
    "$unknownKeywords": [ /* ... */ ]
  }
}
```

(If the validation is false, then the annotation would appear in `droppedAnnotations`, or we could require that it's always there.)

Note that this isn't a meta-schema validation.  `$unknownKeywords` as an annotation would be commenting on the schema itself rather than the instance.

I'm going to add this to my implementation as an option and expose it in the online validator so people can play with it.

**UPDATE** ☝️ is done. Play with it at https://json-everything.net/json-schema, and select the _Unknown keyword annotation_ option on the left.

### Examples

Schema:
```json
{
  "$id": "/base",
  "foo": {
    "not_a_schema": true
  },
  "type": "object",
  "properties": {
    "bar": { "type": "string" }
  }
}
```
Instance:
```json
{"bar": "value"}
```
Result:
```jsonc
{
  "valid": true,
  "evaluationPath": "",
  "schemaLocation": "https://json-everything.net/base#",
  "instanceLocation": "",
  "annotations": {
    "foo": { "not_a_schema": true },
    "properties": [ "bar" ],
    "$unknownKeywords": [ "foo" ]                         // this is the new annotation
  },
  "details": [
    {
      "valid": true,
      "evaluationPath": "/properties/bar",
      "schemaLocation": "https://json-everything.net/base#/properties/bar",
      "instanceLocation": "/bar"
    }
  ]
}
```

**Comment by jdesrosiers** [comment]
I'm glad you included an example, because I wasn't following what you were proposing without it. It's like a virtual annotation keyword. It doesn't actually exist in the schema. It's an interesting idea, but definitely a new behavior, so we shouldn't think carefully about it.

**Comment by Relequestual** [comment]
> The proposed solutions to this were to
> 
> - put additionalProperties: false in the meta-schema
> - disallow unknown keywords entirely
> 
> We've opted for the second approach with a little bit of the first (we'll have one open and one closed meta-schema).
> 
> But what if we just require an annotation that is a list of the unknown keywords?

When you say "just", do you mean "as opposed to the other solutions"?

**Reply by gregsdennis** [reply]
Yes, that is the intent.  Although, disallowing unknown keywords has additional reasoning, not just this.

But most people say that the reason they want a strict mode is to catch typos and other keywords that aren't defined.  This approach gives them that information without having to have a strict mode.

**Reply by Relequestual** [reply]
I'm not sure I see how this would improve anything over the other solutions.

Currently, a strict mode schema could be used for previous and current versions of the specification, and using existing tooling, "errors" in the schema would be picked up and highlighted in IDEs.

If we were to take the proposed solution from this Discussion, new tooling would need to be used to first provide tha correct output (probably not hard), and then to process it into whatever AST is required for reporting back to the IDE interface.

I think there's potential value from reporting unknown keywords in the output format, but I'm not sure what the perceived benefit is? Maybe being able to use just one tool to transform the output into the required AST anywhere?

As a tangent, having a strict mode meta-schema puts the power of "how strict should this schema be?" into the hands of the schema author, while the solution proposed here also provides that power to the validating entity (such as the IDE), and that could open up inconsistencies across the ecosystem, like we see with AJV and its "strict mode" (which is different to the "strict mode" we are discussing, but causing issues nontheless).

**Reply by jdesrosiers** [reply]
> > When you say "just", do you mean "as opposed to the other solutions"?
>
> Yes, that is the intent.

In seems I misunderstood a bit. I think the solution we came up with for schemas is much a better solution when it comes to schemas. However, unknown property detection is a concern for instances as well and the schema solution (prefixing unknown properties with `x-`) doesn't work for instances. This solution could be useful for instances, but I wouldn't back-out what we've decided for unknown keywords in schemas.

**Reply by gregsdennis** [reply]
> However, unknown property detection is a concern for instances as well and the schema solution (prefixing unknown properties with x-) doesn't work for instances.

I don't follow.  The unknown keywords are in the schema.  This would apply to instance validation.  Look at the example again.

**Reply by jdesrosiers** [reply]
Sorry. Let me try to explain better. Although I wouldn't drop what we've already decided in favor of this approach for addressing unknown keywords, there's a very similar problem that could benefit form this approach.

```json
{
  "type": "object",
  "properties": {
    "foo": { "type": "number" }
  }
}
```

```json
{
  "fooo": 42
}
```

Given this schema, the instance is valid, but only accidentally because there's a misspelling of "foo". Tooling could use your proposed sudo-annotation to more easily detect that "fooo" is an unknown property and warn users that they probably meant "foo". 

This was the kind of thing I originally thought you were proposing this to solve. I'm not sure why I thought that because your example was clearly a schema.

**Reply by gregsdennis** [reply]
You can get that annotation already with `additionalProperties: true`.  But yeah, that would be useful.

**Reply by jdesrosiers** [reply]
Ah, yes. I hadn't thought about that because I don't support that annotation. That same approach could be used with schema validation as well without needing to introduce a sudo-annotation, right?

**Reply by gregsdennis** [reply]
If by "schema validation" you mean "by a meta-schema," then, yeah, I suppose it could.  But who actually meta-schema-validates schemas? 😏

**Comment by gregsdennis** [comment]
Another idea would be to update the output format.  @jdesrosiers has casually mentioned (somewhere) that the root of the output could house additional information about an evaluation, things like dialect.  A list of unknown keywords could be added to that instead of being an annotation.

**Comment by gregsdennis** [comment]
The decision to not support unknown keywords has resulted in requiring implementations to refuse to process such schemas, which means this proposal is no longer relevant.  Closing.
